### PR TITLE
Issue 20717: Use request principal for BCL requests, and update client logout selection logic

### PR DESCRIPTION
**Pull request by ayoho**
For #20717

### CODE REVIEW

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops [hunk 1]
```diff
@@ -192,9 +192,9 @@ LOGOUT_TOKEN_ERROR_GETTING_CLAIMS_FROM_ID_TOKEN=CWWKS1643E: The OpenID Connect p
 LOGOUT_TOKEN_ERROR_GETTING_CLAIMS_FROM_ID_TOKEN.explanation=The ID token might be malformed or a claim within the token might have a value that is an unexpected data type. Alternatively, the token might be missing claims that are required or the issuer of the token is not the issuer that is expected.
 LOGOUT_TOKEN_ERROR_GETTING_CLAIMS_FROM_ID_TOKEN.useraction=Ensure the ID token is formatted correctly. See the error that is included in the message for more information.
 
-LOGOUT_TOKEN_ERROR_GETTING_CLIENTS_TO_LOG_OUT=CWWKS1644E: The OpenID Connect provider cannot determine the set of OpenID Connect clients to log out because of the following error: {0}
-LOGOUT_TOKEN_ERROR_GETTING_CLIENTS_TO_LOG_OUT.explanation=The provider failed to retrieve the set of clients that are registered, or the ID token claims might be malformed.
-LOGOUT_TOKEN_ERROR_GETTING_CLIENTS_TO_LOG_OUT.useraction=For more information, see the error included in the message.
+ERROR_RETRIEVING_CLIENT_TO_BUILD_LOGOUT_TOKENS=CWWKS1644W: The {0} OpenID Connect provider encountered the following error when it retrieved the {1} client to build back-channel logout tokens: {2}
```
Comment (ramkumar-k-9286) [inline dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops]: ****NOTE:** CWWKS1644W being a warning message,  it needs to be an error message.**

Reviewed CWWKS1644E: 
The {0} OpenID Connect provider on retrieving the {1} client to build back-channel logout tokens, encountered the following error: {2}

**Comment by ayoho** [comment]
#build

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops [hunk 2]
```diff
@@ -223,10 +223,18 @@ BACKCHANNEL_LOGOUT_REQUEST_EXCEPTION.useraction=Ensure that the back-channel log
 # NOTE: Start of new NLS prefix block: CWWKS1950 - CWWKS1999
 #
 
-BACKCHANNEL_LOGOUT_REQUESTS_INVOKE_ALL_FAILED=CWWKS1950E: The {0} OpenID Connect provider encountered an error while submitting back-channel logout requests: {1}
-BACKCHANNEL_LOGOUT_REQUESTS_INVOKE_ALL_FAILED.explanation=Sending requests to the back-channel logout URIs was interrupted, or another error occurred while scheduling the requests for execution.
+BACKCHANNEL_LOGOUT_REQUESTS_INVOKE_ALL_FAILED=CWWKS1950E: The {0} OpenID Connect provider encountered an error on submitting back-channel logout requests: {1}
```
Comment (ramkumar-k-9286) [inline dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops:226]: Reviewed CWWKS1950E:
The {0} OpenID Connect provider on submitting back-channel logout requests, encountered the following error: {1}

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops [hunk 3]
```diff
@@ -223,10 +223,18 @@ BACKCHANNEL_LOGOUT_REQUEST_EXCEPTION.useraction=Ensure that the back-channel log
...
+BACKCHANNEL_LOGOUT_REQUESTS_INVOKE_ALL_FAILED.explanation=Requests sent to the back-channel logout URIs were interrupted, or another error occurred in scheduling the requests for execution.
```
Comment (ramkumar-k-9286) [inline dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops]: Reviewed explanation:
Requests sent to the back-channel logout URIs were interrupted, or another error occurred in scheduling the requests to run.

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops [hunk 4]
```diff
@@ -223,10 +223,18 @@ BACKCHANNEL_LOGOUT_REQUEST_EXCEPTION.useraction=Ensure that the back-channel log
...
 BACKCHANNEL_LOGOUT_REQUESTS_INVOKE_ALL_FAILED.useraction=Send another logout request to retry the back-channel logout. For more information, see the error included in the message.
 
 ERROR_BUILDING_OR_SENDING_BACKCHANNEL_LOGOUT_REQUEST=CWWKS1951E: The {0} OpenID Connect provider encountered an error while building or sending a back-channel logout request to {1}: {2}
 ERROR_BUILDING_OR_SENDING_BACKCHANNEL_LOGOUT_REQUEST.explanation=The HTTP connection encountered an error, or the connection failed.
 ERROR_BUILDING_OR_SENDING_BACKCHANNEL_LOGOUT_REQUEST.useraction=Send another logout request to retry the back-channel logout. For more information, see the error included in the message.
+
+ERROR_BUILDING_LOGOUT_TOKEN_BASED_ON_ID_TOKEN=CWWKS1952W: A back-channel logout token that is based on a cached ID token for the {0} client cannot be created due to the following error: {1}
```
Comment (ramkumar-k-9286) [inline dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops]: **NOTE: CWWKS1952W being a warning message, it needs to be an error message.**

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops [hunk 5]
```diff
@@ -223,10 +223,18 @@ BACKCHANNEL_LOGOUT_REQUEST_EXCEPTION.useraction=Ensure that the back-channel log
...
+ERROR_BUILDING_LOGOUT_TOKEN_BASED_ON_ID_TOKEN.explanation=An error occurred in building the logout token string. Alternatively, the issuer of the cached ID token might not match this OpenID Connect provider, or the cached ID token might be missing some claims that are required.
```
Comment (ramkumar-k-9286) [inline dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops:235]: An error occurred in building the logout token string. Alternatively, the issuer of the cached ID token might not match this OpenID Connect provider, or the cached ID token might be missing some claims that are required.

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops [hunk 6]
```diff
@@ -223,10 +223,18 @@ BACKCHANNEL_LOGOUT_REQUEST_EXCEPTION.useraction=Ensure that the back-channel log
...
+ERROR_BUILDING_LOGOUT_TOKEN_BASED_ON_ID_TOKEN.useraction=For more information, see the error included in the message.
+
+ERROR_BUILDING_LOGOUT_TOKEN_BASED_ON_ID_TOKEN_CLAIMS=CWWKS1953E: A back-channel logout token that is based on a cached ID token with claims [{0}] for the {1} client cannot be created due to the following error: {2}
+ERROR_BUILDING_LOGOUT_TOKEN_BASED_ON_ID_TOKEN_CLAIMS.explanation=An error occurred while building the logout token string. Alternatively, the issuer of the cached ID token might not match this OpenID Connect provider, or the cached ID token might be missing some claims that are required.
```
Comment (ramkumar-k-9286) [inline dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops]: explanation:
An error occurred in building the logout token string. Alternatively, the issuer of the cached ID token might not match this OpenID Connect provider, or the cached ID token might be missing some claims that are required.

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops (hunk 1 above)
Comment (ayoho) [reply dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops]: I think the suggested change is more clunky and more difficult to translate than the message as it is. ID style in the past has always seemed to steer towards, "The &lt;thing doing the processing> encountered an error/the following error while &lt;doing something>: &lt;the error>". While the suggestion is valid English, it's a more confusing structure to the sentence.

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops (hunk 2 above)
Comment (ayoho) [reply dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops:226]: Similar comment as before.

File: dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops (hunk 5 above)
Comment (ayoho) [reply dev/com.ibm.ws.security.openidconnect.server/resources/com/ibm/ws/security/openidconnect/server/internal/resources/OidcServerMessages.nlsprops:235]: What you have suggested is identical to what is currently there.

**Comment by skasund** [comment]
L2 message review completed

**Comment by ayoho** [comment]
#libby
#build

**Review by teddyjtorres — APPROVED** [review]

**Pull request merged**